libcfnet/protocol.c

```diff
@@ -88,3 +89,88 @@ Seq *ProtocolOpenDir(AgentConnection *conn, const char *path)
 
     return seq;
 }
+
+bool ProtocolGet(AgentConnection *conn, const char *remote_path,
+                 const char *local_path, const uint32_t file_size, int perms)
+{
+    assert(conn != NULL);
+    assert(remote_path != NULL);
+    assert(local_path != NULL);
+    assert(file_size != 0);
+
+    perms = (perms == 0) ? CF_PERMS_DEFAULT : perms;
+
+    unlink(local_path);
+    FILE *file_ptr = safe_fopen_create_perms(local_path, "w", perms);
+    if (file_ptr == NULL)
+    {
+        return false;
+    }
+
+    char buf[CF_MSGSIZE] = {0};
+    int to_send = snprintf(buf, CF_MSGSIZE, "GET %d %s",
+                           file_size, remote_path);
+    if (to_send <= 0)
+    {
+        unlink(local_path);
+        fclose(file_ptr);
+        return false;
+    }
+
+    int ret = SendTransaction(conn->conn_info, buf, to_send, CF_DONE);
+    if (ret == -1)
+    {
+        return false;
+    }
+
+    uint32_t to_recv = 0;
+    while (to_recv < file_size)
+    {
+        int len = TLSRecv(conn->conn_info->ssl, buf, CF_MSGSIZE);
+        if (len == -1)
+        {
+            break;
+        }
+        else if (len > CF_MSGSIZE)
+        {
+            fprintf(stderr, "What?\n");
+            return false;
+        }
+
+        if (BadProtoReply(buf))
+        {
+            Log(LOG_LEVEL_ERR, "Protocol error: %s", buf);
+            unlink(local_path);
+            fclose(file_ptr);
+            return false;
+        }
+
+        ret = fwrite(buf, sizeof(char), len, file_ptr);
+        if (ret < 0)
+        {
+            fprintf(stderr, "bad\n");
+            unlink(local_path);
+            fclose(file_ptr);
+            return false;
```

Comment:
generally, do you want to pay attention to return value/errors on unlink(), fclose(), etc? There are many system calls here and I know it would get messy if each is handled. I guess the question is what can you do if one errors and is it significant to the code that follows?

Reply:
If it fails to unlink, there isn't really much we can do apart from trying again, but stalling the program until the file is successfully deleted (for instance if the file is busy, like a mount point or something) is a bad solution. We could notify the user that the file could not be deleted successfully?

The same goes for `fclose`. I don't know the implications of if a file descriptor remains open in a program until it quits, but `fclose` failing will mean memory leaks since the `FILE` struct won't be deallocated.